src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs

```diff
@@ -45,60 +57,224 @@ internal partial class CommonReferenceManager<TCompilation, TAssemblySymbol>
         /// -    Result of resolving assembly references of the corresponding assembly 
         ///     against provided set of assembly definitions. Essentially, this is an array returned by
         ///     <see cref="AssemblyData.BindAssemblyReferences(ImmutableArray{AssemblyData}, AssemblyIdentityComparer)"/> method.
-        /// </returns>
+        /// </return>
         internal BoundInputAssembly[] Bind(
-            ImmutableArray<AssemblyData> assemblies,
+            ref ImmutableArray<AssemblyData> assemblies,
+            MetadataReferenceResolver resolverOpt,
+            MetadataImportOptions importOptions,
+            ArrayBuilder<MetadataReference> implicitlyResolvedReferences,
+            DiagnosticBag resolutionDiagnostics,
             out bool hasCircularReference,
             out int corLibraryIndex)
         {
             Debug.Assert(!assemblies.IsDefault);
 
-            int totalAssemblies = assemblies.Length;
+            var referenceBindings = ArrayBuilder<AssemblyReferenceBinding[]>.GetInstance();
+            try
+            {
+                // Based on assembly identity, for each assembly, 
+                // bind its references against the other assemblies we have.
+                for (int i = 0; i < assemblies.Length; i++)
+                {
+                    referenceBindings.Add(assemblies[i].BindAssemblyReferences(assemblies, IdentityComparer));
+                }
+
+                if (resolverOpt?.ResolveMissingAssemblies == true)
+                {
+                    ResolveAndBindMissingAssemblies(assemblies, referenceBindings, implicitlyResolvedReferences, resolverOpt, importOptions, resolutionDiagnostics, out assemblies);
+                }
+
+                hasCircularReference = CheckCircularReference(referenceBindings);
+
+                corLibraryIndex = IndexOfCorLibrary(assemblies);
+
+                // For each assembly, locate AssemblySymbol with similar reference resolution
+                // What does similar mean?
+                // Similar means: 
+                // 1) The same references are resolved against the assemblies that we are given.
+                // 2) The same assembly is used as the COR library.
+
+                var boundInputs = new BoundInputAssembly[referenceBindings.Count];
```

Comment:
Are referenceBindings and assemblies supposed to have the same number of elements at this point? Consider asserting if this is a valid expectation.